File: server/lib/phonicsMastery.js
// What it means to KNOW a phonics element, expressed as a pure function.
//
// This is the whole claim behind Dragon Phonics' progress reporting, so it is
// deliberately a module with no database and no imports: the rule can be read,
// argued with, and tested on plain arrays (server/lib/phonicsMastery.test.js).
// The route in server/routes/phonics.js does the querying and hands rows here.
//
// THE RULE, AND WHY IT IS SHAPED THIS WAY.
//
// 1. Only the RECENT window counts. A child who missed /sh/ twenty times in
//    March and has since got it right every time knows /sh/; a lifetime average
//    would say otherwise for months. So each element is judged on its last
//    `RECENT_WINDOW` attempts and old attempts fall out of the verdict — they
//    stay in the table for the trend chart, they just stop voting.
//
// 2. Mastery requires MORE THAN ONE GAME MODE. This is the part that makes
//    "comprehensively understands phonics" mean something. Picking /br/ from
//    four tiles is recognition; typing `br` from the sound alone is recall; and
//    hearing "brick" and finding /br/ inside it is analysis. A child can look
//    fluent on the multiple-choice game by elimination, so a high score in one
//    mode tops out at `solid` — `mastered` is only reachable by being right in
//    at least `MODES_FOR_MASTERY` different modes.
//
// 3. Accuracy alone is not enough — there is a floor on how many attempts a
//    verdict may rest on. Three-for-three is a good day, not proof.
//
// 4. Mastery goes STALE rather than being revoked. Decoding decays without
//    practice, but a child who has not seen /oi/ in two months has not got
//    worse at it — we simply no longer know. `stale` says "re-check this",
//    which is what drives the review round, and it never silently rewrites a
//    level the child earned.

// Attempts considered per element. Ten is roughly one round's worth of an
// element in the mixed-review game, so a bad day is recoverable within a day.
const RECENT_WINDOW = 10;

const MIN_ATTEMPTS_SOLID = 4;
const MIN_ATTEMPTS_MASTERED = 6;
const SOLID_ACCURACY = 0.75;
const MASTERED_ACCURACY = 0.9;
const MODES_FOR_MASTERY = 2;

// After this long unpracticed, a solid/mastered element is flagged for review.
const STALE_AFTER_DAYS = 45;

// Ranked worst to best. The UI colours tiles by this, and `LEVELS.indexOf` is
// how "did this child improve?" is compared, so the order is load-bearing.
const LEVELS = ['new', 'learning', 'solid', 'mastered'];

// Separator for the confusion-pair tally key. A tab can never appear in an
// element key (they are ASCII letters and hyphens), so it cannot collide.
const PAIR_SEP = '\t';

function toDate(value) {
  if (value instanceof Date) return Number.isNaN(value.getTime()) ? null : value;
  if (typeof value === 'string' || typeof value === 'number') {
    const d = new Date(value);
    return Number.isNaN(d.getTime()) ? null : d;
  }
  return null;
}

/**
 * Judge one element from its attempt history.
 *
 * @param {Array<{mode: string, correct: boolean, createdAt: Date|string}>} attempts
 *        Newest first. Extra fields are ignored.
 * @param {Date} [now]  injected so tests can pin "today".
 * @returns {{level: string, attempts: number, correct: number, accuracy: number|null,
 *            modes: string[], lastSeenAt: string|null, stale: boolean, total: number}}
 */
function classifyElement(attempts, now = new Date()) {
  const all = Array.isArray(attempts) ? attempts : [];
  const recent = all.slice(0, RECENT_WINDOW);
  const total = all.length;

  if (recent.length === 0) {
    return {
      level: 'new',
      attempts: 0,
      correct: 0,
      accuracy: null,
      modes: [],
      lastSeenAt: null,
      stale: false,
      total: 0,
    };
  }

  const correct = recent.filter((a) => a.correct).length;
  const accuracy = correct / recent.length;

  // Modes only count when the child was RIGHT in them: being wrong in three
  // different games is not three kinds of evidence that they know it.
  const modes = [...new Set(recent.filter((a) => a.correct).map((a) => a.mode))].sort();

  let level = 'learning';
  if (
    recent.length >= MIN_ATTEMPTS_MASTERED
    && accuracy >= MASTERED_ACCURACY
    && modes.length >= MODES_FOR_MASTERY
  ) {
    level = 'mastered';
  } else if (recent.length >= MIN_ATTEMPTS_SOLID && accuracy >= SOLID_ACCURACY) {
    level = 'solid';
  }

  const lastSeen = toDate(recent[0].createdAt);
  const stale = level !== 'learning'
    && lastSeen != null
    && (now.getTime() - lastSeen.getTime()) > STALE_AFTER_DAYS * 86400000;

  return {
    level,
    attempts: recent.length,
    correct,
    accuracy,
    modes,
    lastSeenAt: lastSeen ? lastSeen.toISOString() : null,
    stale,
    total,
  };
}

/**
 * Judge every element a child has attempted.
 *
 * Elements the child has NEVER attempted are deliberately absent rather than
 * seeded as `new`: the curriculum lives in the frontend (src/data/
 * phonicsCurriculum.js), so the server does not know the full element list and
 * must not pretend to. The client fills the gaps — which also means adding an
 * element to the curriculum does not need a server deploy.
 *
 * @param {Array<{elementKey: string, mode: string, correct: boolean, createdAt: Date|string}>} rows
 *        Any order; grouped and sorted here.
 * @param {Date} [now]
 * @returns {Record<string, ReturnType<typeof classifyElement>>}
 */
function classifyAll(rows, now = new Date()) {
  const byElement = new Map();
  for (const row of rows || []) {
    const key = row.elementKey;
    if (!key) continue;
    if (!byElement.has(key)) byElement.set(key, []);
    byElement.get(key).push(row);
  }

  const out = {};
  for (const [key, attempts] of byElement) {
    attempts.sort((a, b) => {
      const at = toDate(a.createdAt)?.getTime() ?? 0;
      const bt = toDate(b.createdAt)?.getTime() ?? 0;
      return bt - at; // newest first
    });
    out[key] = classifyElement(attempts, now);
  }
  return out;
}

/**
 * The pairs a child actually mixes up, strongest first.
 *
 * A wrong answer only lands here when the child CHOSE a specific other element
 * (`chosen`), which the multiple-choice and find-in-word games always record and
 * the type-it game records when what they typed matches another element's
 * spelling. That is the difference between "gets /sh/ wrong" and "reads /sh/ as
 * /ch/" — only the second one tells a grown-up what to teach.
 *
 * @param {Array<{elementKey: string, chosen: string|null, correct: boolean}>} rows
 * @param {number} [limit]
 * @returns {Array<{element: string, chose: string, count: number}>}
 */
function confusionPairs(rows, limit = 8) {
  const counts = new Map();
  for (const row of rows || []) {
    if (row.correct) continue;
    if (!row.elementKey || !row.chosen) continue;
    if (row.chosen === row.elementKey) continue; // not a confusion
    const id = `${row.elementKey}${PAIR_SEP}${row.chosen}`;
    counts.set(id, (counts.get(id) || 0) + 1);
  }
  return [...counts.entries()]
    .map(([id, count]) => {
      const [element, chose] = id.split(PAIR_SEP);
      return { element, chose, count };
    })
    // A single slip is noise; two of the same mistake is a pattern.
    .filter((p) => p.count >= 2)
    .sort((a, b) => b.count - a.count || a.element.localeCompare(b.element))
    .slice(0, limit);
}

module.exports = {
  RECENT_WINDOW,
  MIN_ATTEMPTS_SOLID,
  MIN_ATTEMPTS_MASTERED,
  SOLID_ACCURACY,
  MASTERED_ACCURACY,
  MODES_FOR_MASTERY,
  STALE_AFTER_DAYS,
  LEVELS,
  classifyElement,
  classifyAll,
  confusionPairs,
};

File: ios/Packages/GameRules/Sources/GameRules/PhonicsMastery.swift
// Phonics mastery — the Swift port of server/lib/phonicsMastery.js, pinned by
// the `mastery` cases of golden/phonics.json, plus the Sound Map's rollups
// (`fullMastery`, `stageSummary`, `overallSummary` in
// src/hooks/usePhonicsProgress.js).
//
// The rule, in short (the JavaScript header argues for each part):
//   1. only an element's last `recentWindow` attempts vote;
//   2. `mastered` needs right answers in at least `modesForMastery` games —
//      one game alone tops out at `solid`;
//   3. a verdict needs a floor of attempts (`minAttemptsSolid`/`Mastered`);
//   4. a solid or mastered element unpractised for `staleAfterDays` is
//      `stale`: flagged for review, never demoted.
//
// Times are milliseconds since 1970 (GameRules imports nothing, so no Date):
// nil is the JavaScript's unparseable date — it sorts as 0 and is never stale.

/// How well a child knows one element, worst to best (`LEVELS`). The order
/// is load-bearing: comparing levels is how "did this improve?" is asked.
public enum PhonicsMasteryLevel: String, Sendable, Hashable, CaseIterable, Comparable, Decodable {
    case new
    case learning
    case solid
    case mastered

    public static func < (lhs: Self, rhs: Self) -> Bool {
        allCases.firstIndex(of: lhs)! < allCases.firstIndex(of: rhs)!
    }
}

/// One answer as the mastery rule reads it: a phonics_attempts row.
public struct PhonicsMasteryAttempt: Sendable, Hashable {
    /// The element asked; nil rows are skipped (`classifyAll`).
    public let elementKey: String?
    /// The game: "choose", "type-it", "find-in-word" or "missing-sound".
    public let mode: String
    public let correct: Bool
    /// For a wrong answer, the element it named; nil otherwise.
    public let chosen: String?
    /// When it was answered, in ms since 1970; nil when unknown.
    public let atMs: Double?

    public init(elementKey: String?, mode: String, correct: Bool, chosen: String? = nil, atMs: Double?) {
        self.elementKey = elementKey
        self.mode = mode
        self.correct = correct
        self.chosen = chosen
        self.atMs = atMs
    }
}

/// One element's verdict (`classifyElement`'s result).
public struct PhonicsElementMastery: Sendable, Hashable {
    public let level: PhonicsMasteryLevel
    /// Attempts in the recent window.
    public let attempts: Int
    /// Right answers in the recent window.
    public let correct: Int
    /// `correct / attempts`; nil with no attempts.
    public let accuracy: Double?
    /// The games answered RIGHT in the window, sorted.
    public let modes: [String]
    /// The newest attempt's time; nil with none or when it's unknown.
    public let lastSeenAtMs: Double?
    /// Solid or mastered, but not practised for `staleAfterDays`.
    public let stale: Bool
    /// Every attempt ever, window or not.
    public let total: Int

    public init(
        level: PhonicsMasteryLevel, attempts: Int, correct: Int, accuracy: Double?, modes: [String],
        lastSeenAtMs: Double?, stale: Bool, total: Int
    ) {
        self.level = level
        self.attempts = attempts
        self.correct = correct
        self.accuracy = accuracy
        self.modes = modes
        self.lastSeenAtMs = lastSeenAtMs
        self.stale = stale
        self.total = total
    }

    /// An element never attempted (`EMPTY_STATE`).
    public static let new = PhonicsElementMastery(
        level: .new, attempts: 0, correct: 0, accuracy: nil, modes: [], lastSeenAtMs: nil, stale: false, total: 0)

    /// The fields a round reads (what GET /api/phonics/mastery reports).
    public var state: PhonicsMasteryState {
        PhonicsMasteryState(level: level.rawValue, stale: stale, accuracy: accuracy)
    }
}

/// A pair a child mixes up: asked `element`, answered `chose`, `count` times.
public struct PhonicsConfusion: Sendable, Hashable {
    public let element: String
    public let chose: String
    public let count: Int

    public init(element: String, chose: String, count: Int) {
        self.element = element
        self.chose = chose
        self.count = count
    }
}

public enum PhonicsMastery {
    /// Attempts considered per element (`RECENT_WINDOW`).
    public static let recentWindow = 10
    public static let minAttemptsSolid = 4
    public static let minAttemptsMastered = 6
    public static let solidAccuracy = 0.75
    public static let masteredAccuracy = 0.9
    public static let modesForMastery = 2
    /// After this long unpractised, a solid/mastered element is stale.
    public static let staleAfterDays = 45

    private static let dayMs = 86_400_000.0

    /// Judge one element from its attempts, newest first (`classifyElement`).
    public static func classifyElement(_ attempts: [PhonicsMasteryAttempt], nowMs: Double) -> PhonicsElementMastery {
        let recent = Array(attempts.prefix(recentWindow))
        guard let newest = recent.first else { return .new }

        let correct = recent.filter(\.correct).count
        let accuracy = Double(correct) / Double(recent.count)
        // Only games answered right count: being wrong in three games is not
        // three kinds of evidence.
        let modes = Array(Set(recent.filter(\.correct).map(\.mode))).sorted()

        var level = PhonicsMasteryLevel.learning
        if recent.count >= minAttemptsMastered && accuracy >= masteredAccuracy && modes.count >= modesForMastery {
            level = .mastered
        } else if recent.count >= minAttemptsSolid && accuracy >= solidAccuracy {
            level = .solid
        }

        let lastSeen = newest.atMs
        let stale = level != .learning && lastSeen.map { nowMs - $0 > Double(staleAfterDays) * dayMs } == true

        return PhonicsElementMastery(
            level: level, attempts: recent.count, correct: correct, accuracy: accuracy, modes: modes,
            lastSeenAtMs: lastSeen, stale: stale, total: attempts.count)
    }

    /// Judge every element attempted, from rows in any order (`classifyAll`).
    /// Elements never attempted are absent: `overview` fills them in as new.
    /// Rows with the same time keep their order (the JavaScript sort is stable).
    public static func classifyAll(_ rows: [PhonicsMasteryAttempt], nowMs: Double) -> [String: PhonicsElementMastery] {
        var byElement: [String: [(Int, PhonicsMasteryAttempt)]] = [:]
        for (index, row) in rows.enumerated() {
            guard let key = row.elementKey, !key.isEmpty else { continue }
            byElement[key, default: []].append((index, row))
        }
        var out: [String: PhonicsElementMastery] = [:]
        for (key, attempts) in byElement {
            let newestFirst = attempts.sorted { a, b in
                let at = a.1.atMs ?? 0, bt = b.1.atMs ?? 0
                return at != bt ? at > bt : a.0 < b.0
            }
            out[key] = classifyElement(newestFirst.map(\.1), nowMs: nowMs)
        }
        return out
    }

    /// The pairs a child actually mixes up, strongest first (`confusionPairs`):
    /// wrong answers that named another element, at least twice.
    public static func confusionPairs(_ rows: [PhonicsMasteryAttempt], limit: Int = 8) -> [PhonicsConfusion] {
        var order: [String: Int] = [:]
        var pairs: [PhonicsConfusion] = []
        for row in rows where !row.correct {
            guard let element = row.elementKey, !element.isEmpty, let chose = row.chosen, !chose.isEmpty,
                chose != element
            else { continue }
            let id = "\(element)\t\(chose)"
            if let at = order[id] {
                pairs[at] = PhonicsConfusion(element: element, chose: chose, count: pairs[at].count + 1)
            } else {
                order[id] = pairs.count
                pairs.append(PhonicsConfusion(element: element, chose: chose, count: 1))
            }
        }
        // A single slip is noise; two of the same mistake is a pattern.
        return Array(
            pairs.enumerated()
                .filter { $0.1.count >= 2 }
                .sorted { a, b in
                    if a.1.count != b.1.count { return a.1.count > b.1.count }
                    if a.1.element != b.1.element { return a.1.element < b.1.element }
                    return a.0 < b.0
                }
                .prefix(limit)
                .map(\.1))
    }

    /// The round builder's view of a verdict map.
    public static func states(_ mastery: [String: PhonicsElementMastery]) -> [String: PhonicsMasteryState] {
        mastery.mapValues(\.state)
    }
}

// MARK: - The Sound Map's rollups

/// How many elements sit at each level, and how many are stale.
public struct PhonicsMasteryCounts: Sendable, Hashable {
    public var new = 0
    public var learning = 0
    public var solid = 0
    public var mastered = 0
    public var stale = 0

    public init() {}

    public subscript(level: PhonicsMasteryLevel) -> Int {
        get {
            switch level {
            case .new: new
            case .learning: learning
            case .solid: solid
            case .mastered: mastered
            }
        }
        set {
            switch level {
            case .new: new = newValue
            case .learning: learning = newValue
            case .solid: solid = newValue
            case .mastered: mastered = newValue
            }
        }
    }

    mutating func add(_ state: PhonicsElementMastery) {
        self[state.level] += 1
        if state.stale { stale += 1 }
    }
}

/// One stage's line on the Sound Map and its card in the stage picker
/// (`stageSummary`).
public struct PhonicsStageSummary: Sendable, Hashable, Identifiable {
    public let stage: PhonicsStage
    public let elements: [PhonicsElement]
    public let counts: PhonicsMasteryCounts

    public var id: Int { stage.stage }
    public var total: Int { elements.count }
    public var mastered: Int { counts.mastered }
    /// Mastered only, on purpose: "solid or better" would let a child who has
    /// only played one game show a full bar.
    public var percent: Int { PhonicsMasteryOverview.percent(counts.mastered, of: total) }
    /// Everything met at all.
    public var touched: Int { total - counts.new }
}

/// The whole curriculum with a verdict for every element, including those
/// never attempted (`fullMastery` + `stageSummary` + `overallSummary`).
public struct PhonicsMasteryOverview: Sendable, Hashable {
    /// Every curriculum element's verdict.
    public let elements: [String: PhonicsElementMastery]
    public let stages: [PhonicsStageSummary]
    /// Program-wide.
    public let overall: PhonicsMasteryCounts

    public var total: Int { PhonicsElement.all.count }
    public var percent: Int { Self.percent(overall.mastered, of: total) }

    /// - Parameter mastery: `classifyAll`'s verdicts; missing elements are new.
    public init(_ mastery: [String: PhonicsElementMastery]) {
        var full: [String: PhonicsElementMastery] = [:]
        var overall = PhonicsMasteryCounts()
        for element in PhonicsElement.all {
            let state = mastery[element.key] ?? .new
            full[element.key] = state
            overall.add(state)
        }
        elements = full
        self.overall = overall
        stages = PhonicsStage.all.map { stage in
            let members = stage.elements
            var counts = PhonicsMasteryCounts()
            for element in members { counts.add(full[element.key] ?? .new) }
            return PhonicsStageSummary(stage: stage, elements: members, counts: counts)
        }
    }

    public subscript(key: String) -> PhonicsElementMastery { elements[key] ?? .new }

    /// `Math.round(part / whole * 100)`, 0 for an empty whole.
    static func percent(_ part: Int, of whole: Int) -> Int {
        whole == 0 ? 0 : Int((Double(part) / Double(whole) * 100).rounded())
    }
}
